versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -1116,7 +1116,31 @@ protected void checkForModifiedKeysBetweenExpectedAndCurrentCommit(
     Set<ContentId> contentIdsForGlobal = new HashSet<>();
     try (Stream<CommitLogEntry> log =
         takeUntilExcludeLast(readCommitLogStream(ctx, refHead), e -> remainingKeys.isEmpty())) {
-      log.peek(entry -> entry.getDeletes().forEach(remainingKeys::remove))
+      log.peek(
+              entry -> {
+                // remove deleted keys from keys to look for
+                entry.getDeletes().forEach(remainingKeys::remove);
+
+                if (entry.getKeyList() != null) {
```

Comment:
Possible extra optimization: I suppose once we process one full key list entry, we do not have to process full key lists again. For example when we are searching for an _exisiting_ key, whose entry is behind _several_ entries with full key lists.

Reply:
Also, I wonder if it might be valuable to add some unreachable history to our correctness tests... i.e. first run tests with empty history, then run them again with N commits add and removing keys, but leaving no "active" keys in HEAD.... WDYT?

Reply:
> Possible extra optimization: I suppose once we process one full key list entry, we do not have to process full key lists again. For example when we are searching for an _exisiting_ key, whose entry is behind _several_ entries with full key lists.

That's tackled by #3592

Reply:
> Also, I wonder if it might be valuable to add some unreachable history to our correctness tests... i.e. first run tests with empty history, then run them again with N commits adding and removing keys, but leaving no "active" keys in HEAD.... WDYT? (a separate PR for sure)

Need to add tests for #3592 incl upgrade tests anyway, so I can put that into that PR.